# Task 4


In [1]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from gsp import *

**Loading the dataset**

For this task we use the original dataset, after having cleaning it. In fact we are now importing the dataset exported from the Task 1 notebook, after assessing and improving data quality (handling missing values and outliers).

In [3]:
df = pd.read_csv('dataset/df.csv', sep=',', index_col=0)

In [4]:
df.head()

BasketID           BasketDate  Sale  CustomerID CustomerCountry  ProdID  \
0   536365  2010-01-12 08:26:00  2.55     17850.0  United Kingdom  85123A   
1   536365  2010-01-12 08:26:00  3.39     17850.0  United Kingdom   71053   
2   536365  2010-01-12 08:26:00  2.75     17850.0  United Kingdom  84406B   
3   536365  2010-01-12 08:26:00  3.39     17850.0  United Kingdom  84029G   
4   536365  2010-01-12 08:26:00  3.39     17850.0  United Kingdom  84029E   

                             ProdDescr  Qta  Sale_per_Qta  
0   WHITE HANGING HEART T-LIGHT HOLDER    6         15.30  
1                  WHITE METAL LANTERN    6         20.34  
2       CREAM CUPID HEARTS COAT HANGER    8         22.00  
3  KNITTED UNION FLAG HOT WATER BOTTLE    6         20.34  
4       RED WOOLLY HOTTIE WHITE HEART.    6         20.34

## Modelling sequences

Now we model each customer as a sequence of baskets.

In [5]:
baskets = df.groupby('CustomerID')['BasketID'].unique().apply(list)
baskets.head()

CustomerID
12347.0    [537626, 542237, 549222, 556201, 562032, 57351...
12348.0                             [539318, 541998, 548955]
12349.0                                             [577609]
12350.0                                             [543037]
12352.0    [544156, 545323, 546869, 547390, 567505, 56869...
Name: BasketID, dtype: object

We drop the customers having performed only one shopping session.

In [6]:
min_baskets = 2

In [7]:
customers_to_drop = baskets[baskets.apply(len) < min_baskets].index

In [8]:
baskets.drop(index=customers_to_drop, inplace=True)

We compute now a dataframe where each row presents a basket ID and the products bought during said transaction.

In [9]:
transactions = df[df.CustomerID.isin(baskets.index)].groupby('BasketID')['ProdID'].unique().apply(list)
transactions.head()

BasketID
536365       [85123A, 71053, 84406B, 84029G, 84029E, 21730]
536366                                              [22633]
536367    [22745, 22748, 22749, 22310, 84969, 22623, 217...
536368                         [22960, 22913, 22912, 22914]
536369                                              [21756]
Name: ProdID, dtype: object

We now combine the two dataframe in order to compute a list of each product bought by each customer during each of his sessions.

In [10]:
df2 = pd.DataFrame(columns=['CustomerID', 'basket_list'])
for CustomerID in baskets.index:
    customer_list = []
    for BasketID in baskets.loc[CustomerID]:
        customer_list.append(transactions.loc[BasketID])
    current_customer = pd.DataFrame(data= {'CustomerID': CustomerID, 'basket_list': [customer_list]})
    df2 = pd.concat([df2, current_customer])

In [11]:
df2.set_index('CustomerID', inplace=True)

This is the resulting dataframe.

In [12]:
df2

basket_list
CustomerID                                                   
12347.0     [[85116, 22375, 71477, 22771, 22772, 22773, 22...
12348.0      [[84991, 21213, 22952, 21977], [21726], [22437]]
12352.0     [[21380, 22064, 21232, 22646, 22779, 22654, 21...
12356.0     [[22138, 22062, 22066, 22131, 22195, 22937, 84...
12358.0     [[15060B, 22059, 37447, 15056P, 15056N, 20679,...
...                                                       ...
18273.0                                  [[79302M], [79302M]]
18282.0     [[23295, 22089, 21108, 21109], [22699, 22818, ...
18283.0     [[22356, 20726, 22384, 22386, 20717, 20718, 85...
18287.0     [[22755, 22754, 22753, 22756, 22758, 22757, 22...
18288.0     [[85183A, 21058, 21967, 21981, 21246, 84997D, ...

[2720 rows x 1 columns]

We now format slighty modify the shape of the dataframe so to fit the requirements for using it as an input of the apriori function.

In [13]:
dataset = []
for row in df2.values:
    dataset.append(row[0])

We count the total number of sequences and events contained within the dataset.

In [15]:
print("Number of input sequences: ", len(dataset))
print("Total number of events: ", sum([len(seq) for seq in dataset])) 

Number of input sequences:  2720
Total number of events:  17378


We compute the support representing particular percentage of the dataset population (from 5% to 50%)

In [17]:
for y in [5, 10, 15, 20, 40, 50]:
    print((len(dataset)*y) / 100.0)

136.0
272.0
408.0
544.0
1088.0
1360.0


We start by computing the frequent patterns with minimum support equal to the 5% of the dataset population.

In [18]:
itemsets_5 = apriori(dataset, minSupport=136, verbose=False) 

In [26]:
itemsets_5

[([['15036']], 137),
 ([['15056N']], 142),
 ([['16161P']], 157),
 ([['20676']], 154),
 ([['20711']], 136),
 ([['20712']], 208),
 ([['20713']], 185),
 ([['20718']], 199),
 ([['20719']], 202),
 ([['20723']], 191),
 ([['20724']], 276),
 ([['20725']], 465),
 ([['20726']], 336),
 ([['20727']], 410),
 ([['20728']], 411),
 ([['20914']], 288),
 ([['20969']], 140),
 ([['20971']], 233),
 ([['20972']], 261),
 ([['20973']], 168),
 ([['20974']], 165),
 ([['20975']], 196),
 ([['20981']], 137),
 ([['20983']], 152),
 ([['21034']], 373),
 ([['21080']], 360),
 ([['21086']], 146),
 ([['21094']], 152),
 ([['21121']], 198),
 ([['21122']], 178),
 ([['21124']], 152),
 ([['21136']], 198),
 ([['21154']], 146),
 ([['21155']], 179),
 ([['21156']], 179),
 ([['21164']], 144),
 ([['21165']], 150),
 ([['21166']], 243),
 ([['21169']], 162),
 ([['21172']], 204),
 ([['21174']], 192),
 ([['21175']], 286),
 ([['21181']], 295),
 ([['21210']], 203),
 ([['21212']], 520),
 ([['21213']], 238),
 ([['21231']], 168),
 ([['21232'

We compute the frequent patterns with minimum support equal to the 10% of the dataset population.

In [19]:
itemsets_10 = apriori(dataset, minSupport=272, verbose=False)

In [27]:
itemsets_10

[([['20724']], 276),
 ([['20725']], 465),
 ([['20726']], 336),
 ([['20727']], 410),
 ([['20728']], 411),
 ([['20914']], 288),
 ([['21034']], 373),
 ([['21080']], 360),
 ([['21175']], 286),
 ([['21181']], 295),
 ([['21212']], 520),
 ([['21485']], 321),
 ([['21733']], 309),
 ([['21754']], 324),
 ([['21755']], 296),
 ([['21790']], 365),
 ([['21791']], 282),
 ([['21889']], 282),
 ([['21915']], 283),
 ([['21931']], 282),
 ([['21977']], 327),
 ([['22077']], 335),
 ([['22086']], 476),
 ([['22111']], 352),
 ([['22112']], 317),
 ([['22114']], 280),
 ([['22138']], 460),
 ([['22139']], 405),
 ([['22144']], 272),
 ([['22149']], 282),
 ([['22178']], 319),
 ([['22197']], 308),
 ([['22382']], 426),
 ([['22383']], 378),
 ([['22384']], 395),
 ([['22386']], 306),
 ([['22411']], 312),
 ([['22457']], 487),
 ([['22469']], 462),
 ([['22470']], 407),
 ([['22551']], 286),
 ([['22554']], 299),
 ([['22558']], 293),
 ([['22621']], 300),
 ([['22629']], 283),
 ([['22662']], 276),
 ([['22666']], 409),
 ([['22697']]

We compute the frequent patterns with minimum support equal to the 15% of the dataset population.

In [20]:
itemsets_15 = apriori(dataset, minSupport=408, verbose=False)

In [28]:
itemsets_15

[([['20725']], 465),
 ([['20727']], 410),
 ([['20728']], 411),
 ([['21212']], 520),
 ([['22086']], 476),
 ([['22138']], 460),
 ([['22382']], 426),
 ([['22457']], 487),
 ([['22469']], 462),
 ([['22666']], 409),
 ([['22720']], 528),
 ([['22960']], 462),
 ([['22961']], 434),
 ([['23203']], 432),
 ([['23245']], 416),
 ([['23298']], 508),
 ([['47566']], 570),
 ([['84879']], 517),
 ([['85099B']], 506),
 ([['85123A']], 624)]

We compute the frequent patterns with minimum support equal to the 20% of the dataset population.

In [21]:
itemsets_20 = apriori(dataset, minSupport=544, verbose=False)

In [29]:
itemsets_20

[([['47566']], 570), ([['85123A']], 624)]

The searches for pattern with a minimum support equal to the 25% and 30% didn't give any results and as such are included only as comments.

In [ ]:
# itemsets_25 = apriori(dataset, minSupport=680, verbose=False)

In [ ]:
# itemsets_30 = apriori(dataset, minSupport=816, verbose=False)

We print how many patterns have been identified for each percentage of support.

In [24]:
print('5% of support: ', str(len(itemsets_5)))
print('10% of support: ' , str(len(itemsets_10)))
print('15% of support: ' , str(len(itemsets_15)))
print('20% of support: ' , str(len(itemsets_20)))
#print('25% of support: ' str(len(itemsets_25)))
#print('30% of support: ' str(len(itemsets_30)))

5% of support:  749
10% of support:  94
15% of support:  20
20% of support:  2


We now proceed by substituting each ID associated with the products in the aforementioned patterns with its respective description from the original dataset.

In [30]:
def get_description(productID):
    return df[df.ProdID == productID].ProdDescr.values[0]

def itemset_to_descriptions(itemset):
    itemset_with_descriptions = []
    for sequence in itemset:
        
        set_of_products = sequence[0]
        support = sequence[1]
        
        pattern = []
        set_of_descriptions = []
        
        for products in set_of_products:
            description_sequences = []
            
            for product in products:
                product_descr = get_description(product)
                description_sequences.append(product_descr)
            
            set_of_descriptions.append(description_sequences)
        
        pattern = set_of_descriptions, support
        itemset_with_descriptions.append(pattern)
        
    return itemset_with_descriptions

In order to present an example of analysis of the patterns just obtained without slowing down too much the computation time of the notebook in question, the conversion and discussion of the analyzed patterns will involve only those relating to 5% of the dataset population.

In [33]:
itemsets_5_descr = itemset_to_descriptions(itemsets_5)

This are the now converted itemsets.

In [34]:
itemsets_5_descr

[([['ASSORTED COLOURS SILK FAN']], 137),
 ([['EDWARDIAN PARASOL NATURAL']], 142),
 ([['WRAP ENGLISH ROSE ']], 157),
 ([['RED RETROSPOT BOWL']], 154),
 ([['JUMBO BAG TOYS ']], 136),
 ([['JUMBO BAG WOODLAND ANIMALS']], 208),
 ([['JUMBO BAG OWLS']], 185),
 ([['RED RETROSPOT SHOPPER BAG']], 199),
 ([['WOODLAND CHARLOTTE BAG']], 202),
 ([['STRAWBERRY CHARLOTTE BAG']], 191),
 ([['RED RETROSPOT CHARLOTTE BAG']], 276),
 ([['LUNCH BAG RED RETROSPOT']], 465),
 ([['LUNCH BAG WOODLAND']], 336),
 ([['LUNCH BAG  BLACK SKULL.']], 410),
 ([['LUNCH BAG CARS BLUE']], 411),
 ([['SET/5 RED RETROSPOT LID GLASS BOWLS']], 288),
 ([['RED FLORAL FELTCRAFT SHOULDER BAG']], 140),
 ([['PINK BLUE FELT CRAFT TRINKET BOX']], 233),
 ([['PINK CREAM FELT CRAFT TRINKET BOX ']], 261),
 ([['12 PENCIL SMALL TUBE WOODLAND']], 168),
 ([['12 PENCILS SMALL TUBE SKULL']], 165),
 ([['12 PENCILS SMALL TUBE RED RETROSPOT']], 196),
 ([['12 PENCILS TALL TUBE WOODLAND']], 137),
 ([['12 PENCILS TALL TUBE RED RETROSPOT']], 152),
 ([['R

As demonstrated by the very large presence of sequences composed of the same or similar articles (see examples in the next cell), we suspect the correlation of this phenomenon to a habitual purchase of the products in question. 

Examples of sequences with similar or identical products:
- ([['LUNCH BAG RED RETROSPOT', 'LUNCH BAG WOODLAND']], 197),
- ([['ALARM CLOCK BAKELIKE GREEN'], ['ALARM CLOCK BAKELIKE GREEN']], 154)

Possible explanations for this trend are:
- the "disposable" nature of the products in question and therefore the need for customers to repurchase them continuously over time;
- the nature of wholesalers or resellers of the customers themselves who could therefore repeatedly purchase the products in question from the commercial entity under analysis in order to resell them to any third parties.

### Task 4: With time costraints

Here we make some attempts to tackle the optional task of including time constraints within the identification of general sequential patterns.

The previous structure is modified in order to associate each sequence of products with the date of the cart that includes them.

In [4]:
df['BasketDate'] = pd.to_datetime(df['BasketDate']) # converting 'BasketDate' to datetime
df['BasketDate'] = df['BasketDate'].dt.date # cutting the time from to BasketDate

In [56]:
df.BasketDate

0         2010-01-12
1         2010-01-12
2         2010-01-12
3         2010-01-12
4         2010-01-12
             ...    
541904    2011-09-12
541905    2011-09-12
541906    2011-09-12
541907    2011-09-12
541908    2011-09-12
Name: BasketDate, Length: 385516, dtype: object

In [6]:
baskets = df.groupby('CustomerID')['BasketID'].unique().apply(list)
baskets.head()

CustomerID
12347.0    [537626, 542237, 549222, 556201, 562032, 57351...
12348.0                             [539318, 541998, 548955]
12349.0                                             [577609]
12350.0                                             [543037]
12352.0    [544156, 545323, 546869, 547390, 567505, 56869...
Name: BasketID, dtype: object

In order to speed up the computation the customers with less than 5 shopping sessions are excluded from the analysis.

In [7]:
min_baskets = 5

In [8]:
customers_to_drop = baskets[baskets.apply(len) < min_baskets].index

In [9]:
baskets.drop(index=customers_to_drop, inplace=True)

In [11]:
transactions = df[df.CustomerID.isin(baskets.index)].groupby('BasketID')['ProdID'].unique().apply(list)
transactions.head()

BasketID
536365       [85123A, 71053, 84406B, 84029G, 84029E, 21730]
536366                                              [22633]
536367    [22745, 22748, 22749, 22310, 84969, 22623, 217...
536368                         [22960, 22913, 22912, 22914]
536369                                              [21756]
Name: ProdID, dtype: object

In [12]:
df2 = pd.DataFrame(columns=['CustomerID', 'basket_list'])
for CustomerID in baskets.index:
    current_customer_row = []
    for BasketID in baskets.loc[CustomerID]:
        transaction_date = df[df.BasketID == BasketID].BasketDate.unique()[0]
        basket_log = [transaction_date, transactions.loc[BasketID]]
        current_customer_row.append(basket_log)
    current_customer = pd.DataFrame(data= {'CustomerID': CustomerID, 'basket_list': [current_customer_row]})
    df2 = pd.concat([df2, current_customer])

In [13]:
df2.set_index('CustomerID', inplace=True)

This is the new structure accomodating both the sequence of products and the date of their purchase.

In [14]:
df2

basket_list
CustomerID                                                   
12347.0     [[2010-07-12, [85116, 22375, 71477, 22771, 227...
12352.0     [[2011-02-16, [21380, 22064, 21232, 22646, 227...
12362.0     [[2011-02-17, [21913, 22431, 22432, 21746, 217...
12381.0     [[2011-12-08, [47566, 16169E, 21481, 22945, 22...
12383.0     [[2010-12-22, [21121, 21122, 21124, 21125, 211...
...                                                       ...
18257.0     [[2011-02-16, [85099B, 22957, 22502, 22961, 22...
18260.0     [[2010-12-16, [79321, 22452, 22453, 22454, 220...
18272.0     [[2011-07-04, [20754, 21563, 71459, 22557, 229...
18283.0     [[2011-06-01, [22356, 20726, 22384, 22386, 207...
18288.0     [[2011-01-24, [85183A, 21058, 21967, 21981, 21...

[1046 rows x 1 columns]

In [18]:
print("Number of input sequences: ", len(dataset))
print("Total number of events: ", sum([len(seq) for seq in dataset])) 

Number of input sequences:  1046
Total number of events:  12784


Unfortunately the analysis does not continue beyond the definition of the structure due to the proximity of the deadline.